livecheck/heuristic.rb

```diff
@@ -93,52 +63,41 @@ def latest_version(formula)
   urls ||= checkable_urls(formula)
 
   if Homebrew.args.debug?
-    puts "Formula:         #{formula_name(formula)}"
-    puts "Head only?:      #{!formula.stable?}" unless formula.stable?
-    puts "Livecheckable?:  #{has_livecheckable ? "Yes" : "No"}"
+    puts "\nFormula:          #{formula_name(formula)}"
+    puts "Head only?:       #{!formula.stable?}" unless formula.stable?
+    puts "Livecheckable?:   #{has_livecheckable ? "Yes" : "No"}"
   end
 
   urls.each do |original_url|
+    puts "\nURL:              #{original_url}" if Homebrew.args.debug?
+
     # Skip Gists until/unless we create a method of identifying revisions
-    next if original_url.include?("gist.github.com")
+    if original_url.include?("gist.github.com")
+      odebug "Skipping: GitHub Gists are not supported"
+      next
+    end
 
     url = preprocess_url(original_url)
-
-    strategy = if /hackage\.haskell\.org/.match?(url)
-      :hackage_strategy
-    elsif DownloadStrategyDetector.detect(url) <= GitDownloadStrategy
-      :git_strategy
-    elsif /(sourceforge|sf)\.net/.match?(url) &&
-          SOURCEFORGE_SPECIAL_CASES.none? { |sc| url.include? sc }
-      :sourceforge_strategy
-    elsif url =~ /gnu\.org/ && GNU_SPECIAL_CASES.none? { |sc| url.include? sc }
-      :gnu_strategy
-    elsif /files\.pythonhosted\.org/.match?(url)
-      :pypi_strategy
-    elsif /registry\.npmjs\.org/.match?(url)
-      :npm_strategy
-    elsif /download\.gnome\.org/.match?(url)
-      :gnome_strategy
-    elsif /launchpad\.net/.match?(url)
-      :launchpad_strategy
-    elsif %r{www\.apache\.org/dyn}.match?(url)
-      :apache_strategy
-    elsif %r{bitbucket\.org(/[^/]+){4}\.\w+}.match?(url)
-      :bitbucket_strategy
-    elsif livecheck_regex
-      :page_match_strategy
-    end
+    strategies = LivecheckStrategy.from_url(url, livecheck_regex.present?)
+    strategy = strategies[0]
 
     if Homebrew.args.debug?
-      puts "\nURL:             #{original_url}"
-      puts "URL (processed): #{url}" if url != original_url
-      puts "Strategy:        #{strategy.nil? ? "None" : strategy.to_s.delete_suffix("_strategy")}"
-      puts "Regex:           #{livecheck_regex.inspect}\n" unless livecheck_regex.nil?
+      puts "URL (processed):  #{url}" if url != original_url
+      puts "Strategies:       #{strategies.map { |s| s.name.demodulize }}" unless strategies.empty? || !Homebrew.args.verbose?
```

Comment:
Currently it produces output like this:
```
Strategies:       ["PageMatch"]
```
I suggest making it a bit more human-readable:
```suggestion
      puts "Strategies:       #{strategies.map { |s| s.name.demodulize }.join(", ")}" unless strategies.empty? || !Homebrew.args.verbose?
```

Reply:
I can get behind this. I need to make a change to this line to fix a style issue, so I'll incorporate it in a reworking of the latest commit.